pil-analyzer/src/side_effect_checker.rs

```diff
@@ -0,0 +1,124 @@
+use std::collections::HashMap;
+
+use powdr_ast::{
+    analyzed::{
+        Expression, FunctionValueDefinition, Reference, Symbol, SymbolKind, TypedExpression,
+    },
+    parsed::{FunctionKind, LambdaExpression, StatementInsideBlock},
+};
+
+use lazy_static::lazy_static;
+
+/// Check that query functions are only referenced/defined in a query context
+/// and that constr functions are only referenced/defined in a constr context.
+pub fn check(
+    definitions: &HashMap<String, (Symbol, Option<FunctionValueDefinition>)>,
+    context: FunctionKind,
+    e: &Expression,
+) -> Result<(), String> {
+    SideEffectChecker {
+        definitions,
+        context,
+    }
+    .check(e)
+}
+
+struct SideEffectChecker<'a> {
+    definitions: &'a HashMap<String, (Symbol, Option<FunctionValueDefinition>)>,
+    context: FunctionKind,
+}
+
+impl<'a> SideEffectChecker<'a> {
+    fn check(&mut self, e: &Expression) -> Result<(), String> {
+        match e {
+            Expression::Reference(Reference::Poly(r)) => {
+                let kind = self.function_kind_of_symbol(&r.name);
+                if kind != FunctionKind::Pure && kind != self.context {
+                    return Err(format!(
+                        "Referenced a {kind} function inside a {} context: {}",
+                        self.context, r.name,
+                    ));
+                }
+                Ok(())
+            }
+            Expression::LambdaExpression(LambdaExpression {
+                kind,
+                params: _,
+                body,
+            }) => {
+                if *kind != FunctionKind::Pure && *kind != self.context {
+                    return Err(format!(
+                        "Used a {kind} lambda function inside a {} context: {e}",
+                        self.context
+                    ));
+                }
+                let old_context = self.context;
+                let result = self.check(body);
+                self.context = old_context;
+                result
+            }
+            Expression::BlockExpression(statements, _expr) => {
+                for s in statements {
+                    match s {
+                        StatementInsideBlock::LetStatement(s) => {
+                            if s.value.is_none() && self.context != FunctionKind::Constr {
```

Comment:
Wouldn't it be cleaner to check that `s`'s type is `witness`? I guess it's equivalent to `s.value.is_none()` now, just wondering if it may change